### PR TITLE
[data] Re-enable nightly test for MLPerf ImageNet ingest

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Updates MLPerf resnet50 benchmark to use Datasets streaming. Changes:
- migrate to new DatasetConfig APIs
- Update benchmarks - delete heterogeneity benchmark, shorten file size benchmark, separate out OOM benchmark
- Add utils to test Torch - this is not included in the test right now. Ideally we should test it separately with a different codepath using TorchTrainer and iter_torch_batches().
  - `process_imagenet.sh`: File to download 1GB of ImageNet data and restructure directories to expected Torch layout
  - Adds a `--from-images` and `--use-torch` option to `resnet50_ray_air.py`

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by scottjlee — APPROVED** [review]

File: release/air_tests/air_benchmarks/mlperf-train/file_size_benchmark.sh [hunk 1]
```diff
@@ -1,42 +1,63 @@
...
-        | sed 's/ /\n/g' \
-        | python -c "import sys; print(max([int(line) for line in sys.stdin]))")
-NUM_IMAGES_PER_EPOCH=$((MIN_PARALLELISM * MAX_IMAGES_PER_FILE))
-
 SHARD_URL_PREFIX=https://air-example-data.s3.us-west-2.amazonaws.com/air-benchmarks
 
-for num_images_per_file in $NUM_IMAGES_PER_FILE; do
-    num_files=$(python -c "import math; print(math.ceil($NUM_IMAGES_PER_EPOCH/$num_images_per_file))")
-    
+# First epoch is for warmup, report results from second.
```
Comment (scottjlee) [inline release/air_tests/air_benchmarks/mlperf-train/file_size_benchmark.sh:13]: i'm unclear on why we run 2 epochs and report results from the second one. what does it mean for the first epoch to be a "warmup"?

**Comment by stephanie-wang** [comment]
Running here: https://buildkite.com/ray-project/release-tests-pr/builds/45931

File: release/air_tests/air_benchmarks/mlperf-train/file_size_benchmark.sh (hunk 1 above)
Comment (stephanie-wang) [reply release/air_tests/air_benchmarks/mlperf-train/file_size_benchmark.sh:13]: Ah the system might still be warming up on the first epoch, so that's why we're only reporting results for the second.

Comment (stephanie-wang) [reply release/air_tests/air_benchmarks/mlperf-train/file_size_benchmark.sh:13]: (like starting Ray workers, etc)

File: release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py [hunk 2]
```diff
@@ -162,20 +143,18 @@ def to_tensor_iterator():
...
+    for epoch in range(config["num_epochs"]):
         tf_dataset = None
         if config["data_loader"] == TF_DATA:
             assert _tf_dataset is not None
             tf_dataset = _tf_dataset
         elif config["data_loader"] == RAY_DATA:
             assert dataset_shard is not None
-            tf_dataset = ray_dataset_to_tf_dataset(
-                dataset=dataset_shard,
+            tf_dataset = dataset_shard.to_tf(
```
Comment (raulchen) [inline release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py:154]: 1. for comparison with torch, I think we also need to do `iter_torch_batches` when data_loader is RAY_DATA. maybe we should refactor the script structure. 
2. Why not use `iter_tf_batches` instead of `to_tf`? iter should be more efficient?

File: release/air_tests/air_benchmarks/mlperf-train/image_loader_microbenchmark.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+import ray
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply release/air_tests/air_benchmarks/mlperf-train/image_loader_microbenchmark.py]: This was a mistake and has been deleted.

File: release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py (hunk 2 above)
Comment (stephanie-wang) [reply release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py:154]: 1. Yes that's marked as TODO in this code.
2. I think to_tf is the recommended API.

File: release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py [hunk 4]
```diff
@@ -184,45 +162,61 @@ def to_tensor_iterator():
                 num_steps_per_epoch=num_steps_per_epoch,
             )
 
+        epoch_start_time_s = time.perf_counter()
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py:166]: I don't think there is any particular reason (it shouldn't matter).

File: release/air_tests/air_benchmarks/mlperf-train/process_imagenet.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
Comment (raulchen) [inline release/air_tests/air_benchmarks/mlperf-train/process_imagenet.sh:1]: Is this file also duplicated with the one in https://github.com/ray-project/ray/pull/37498/files?

File: release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py (hunk 2 above)
Comment (raulchen) [reply release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py:154]: Could you help me understand why `to_tf` is more preferable? 
It looks to me that for the torch examples, we recommend using `iter_torch_batches`. What is the difference between TF and Torch?

**Review by raulchen — APPROVED** [review]

File: release/air_tests/air_benchmarks/mlperf-train/process_imagenet.sh (hunk 5 above)
Comment (stephanie-wang) [reply release/air_tests/air_benchmarks/mlperf-train/process_imagenet.sh:1]: This one is intentional because we will need it for the torch version of this benchmark that reads directly from images, but it is not used right now in the nightly test.

File: release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py (hunk 2 above)
Comment (stephanie-wang) [reply release/air_tests/air_benchmarks/mlperf-train/resnet50_ray_air.py:154]: The [docs](https://docs.ray.io/en/master/data/api/data_iterator.html) recommend using `iter_torch_batches` for torch and `to_tf` for TF.

I think this was mainly just to reduce friction for users and better match the third-party interfaces for torch DataLoader and tf.data.Dataset. There's some more context [here](https://github.com/ray-project/ray/issues/29031).
